### PR TITLE
Closes 5533 & 5424 Detect use of Perfmatters and RapidLoad used css

### LINKED ISSUE #5424: Detect usage of Autoptimize's RapidLoad Used CSS
**Is your feature request related to a problem? Please describe.**
Currently, customers can enable [RapidLoad Power-Up for Autoptimize]( https://wordpress.org/plugins/unusedcss/) at the same time as WP Rocket's Remove Unused CSS, causing layout issues or stuck Used CSS generation problems.

**Describe the solution you'd like**
WP Rocket could detect the usage of RapidLoad Remove Unused CSS and gray out Remove Unused CSS in WP Rocket, showing a notice similar to [the one shown when Autoptimize's minification is detected](https://d33v4339jhl8k0.cloudfront.net/docs/assets/5415e7bfe4b01e2a68fe8243/images/5ab0e897042863478ea7b0cc/file-LmFLCODRoP.png).

Maybe we can also target any specific Asset CleanUp setting, but I'm not sure.

**Additional context**
Ticket: https://secure.helpscout.net/conversation/1973225238/360828 - RapidLoad was enabled, and the whole layout broke when WP Rocket's Remove Unused CSS was applied.

### LINKED ISSUE #5533: Remove Unused CSS - Detect usage of Perfmatters' Used CSS
**Is your feature request related to a problem? Please describe.**
Whenever Perfmatters' Used CSS is used together with our RUCSS, it leads to the problems with layout.

**Describe the solution you'd like**
WP Rocket could detect the usage of Perfmatter's Used CSS and gray out Remove Unused CSS in WP Rocket, showing a notice similar to [the one shown when Autoptimize's minification is detected](https://d33v4339jhl8k0.cloudfront.net/docs/assets/5415e7bfe4b01e2a68fe8243/images/5ab0e897042863478ea7b0cc/file-LmFLCODRoP.png).

**Additional context**
Should be fixed together with: 
https://github.com/wp-media/wp-rocket/issues/5424

Slack thread: https://wp-media.slack.com/archives/C43T1AYMQ/p1663078682187319 - User was using Perfmatter's Remove Unused CSS, causing WP Rocket's Used CSS generation to get stuck on `failed`.

### PR DESCRIPTION
**Pull request by jeawhanlee**
## Description
This PR detects the use of Perfmatters and RapidLoad used css and disable the rucss option.

Fixes #5533
Fixes #5424 

## Type of change

- [x] Enhancement (non-breaking change which improves an existing functionality)

## Is the solution different from the one proposed during the grooming?
No

## How Has This Been Tested?
Automated Tests

# Checklist:
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Review by CrochetFeve0251 — APPROVED** [review]

**Review by vmanthos — COMMENTED** [review]
This PR is working as expected, i.e. Remove Unused CSS is disabled and a notice is displayed when Perfmatters' Remove Unused CSS or RapidLoad Power-Up for Autoptimize is enabled.

**Notes**

1. In the Perfmatters notice, instead of using `Removing Unused CSS is currently activated in` we should use `Remove Unused CSS is currently activated in` to match the name of their feature.

2. For `RapidLoad Power-Up for Autoptimize` we display this:
> Removing Unused CSS is currently activated in Autoptimize's RapidLoad. If you want to use WP Rocket's Remove Unused CSS feature, disable this option in Autoptimize's RapidLoad.

Taking account the description on RapidLoad Power-Up for Autoptimize [plugin's page](RapidLoad Power-Up for Autoptimize), I suggest we change the notice to:

> Automated unused CSS removal is currently activated in RapidLoad Power-Up for Autoptimize. If you want to use WP Rocket's Remove Unused CSS feature, disable this option in RapidLoad Power-Up for Autoptimize.

3. In the case of RapidLoad Power-Up for Autoptimize, WP Rocket's Remove Unused CSS will be disabled even if someone hasn't actually connected their rapidload.io account, in which case unused CSS still exists.

Ideally, we should disable our feature only when theirs is applied.

**Comment by piotrbak** [comment]
@vmanthos @jeawhanlee Agreed with the wording.

For the 3rd point, @jeawhanlee do we have an easy option to see if their option is **really** activated?

**Comment by jeawhanlee** [comment]
> @vmanthos @jeawhanlee Agreed with the wording.
> 
> For the 3rd point, @jeawhanlee do we have an easy option to see if their option is **really** activated?

@piotrbak I'll check this...

File: inc/Engine/Admin/Settings/Page.php [hunk 1]
```diff
@@ -571,6 +571,19 @@ private function assets_section() {
 		$disable_combine_css = $this->disable_combine_css();
 		$disable_ocd         = 'local' === wp_get_environment_type();
 
+		/**
+		 * Filters the status of the RUCSS option.
+		 *
+		 * @param array $should_disable will return array with disable status and text.
+		 */
+		$rucss_status = apply_filters(
+			'rocket_maybe_disable_rucss',
```
Comment (remyperona) [inline inc/Engine/Admin/Settings/Page.php]:
```suggestion
			'rocket_disable_rucss_setting',
```

File: inc/ThirdParty/Plugins/Optimization/Perfmatters.php [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+	 *
+	 * @return array
+	 */
+	public static function get_subscribed_events() {
+		if ( ! defined( 'PERFMATTERS_VERSION' ) ) {
+			return [];
+		}
+
+		return [
+			'rocket_maybe_disable_rucss'              => 'disable_rucss_option',
```
Comment (remyperona) [inline inc/ThirdParty/Plugins/Optimization/Perfmatters.php]:
```suggestion
			'rocket_disable_rucss_setting'              => 'disable_rucss_setting',
```

File: inc/ThirdParty/Plugins/Optimization/Perfmatters.php [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+		];
+	}
+
+	/**
+	 * Disable RUCSS option.
+	 *
+	 * @param array $status RUCSS option status.
+	 * @return array
+	 */
+	public function disable_rucss_option( array $status ): array {
```
Comment (remyperona) [inline inc/ThirdParty/Plugins/Optimization/Perfmatters.php]:
```suggestion
	public function disable_rucss_setting( array $status ): array {
```

File: inc/ThirdParty/Plugins/Optimization/Perfmatters.php [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+		}
+
+		return [
+			'rocket_maybe_disable_rucss'              => 'disable_rucss_option',
+			'pre_get_rocket_option_remove_unused_css' => 'maybe_disable_rucss',
+		];
+	}
+
+	/**
+	 * Disable RUCSS option.
```
Comment (remyperona) [inline inc/ThirdParty/Plugins/Optimization/Perfmatters.php]:
```suggestion
	 * Disable RUCSS setting.
```

File: inc/ThirdParty/Plugins/Optimization/Perfmatters.php [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+	 * @return bool|null
+	 */
+	public function maybe_disable_rucss() {
+		return $this->is_perfmatters_rucss_active() ? false : null;
+	}
+
+	/**
+	 * Check if perfmatters rucss is active.
+	 *
+	 * @return boolean
```
Comment (remyperona) [inline inc/ThirdParty/Plugins/Optimization/Perfmatters.php]:
```suggestion
	 * @return bool
```

File: inc/ThirdParty/Plugins/Optimization/Perfmatters.php [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+	 */
+	private function is_perfmatters_rucss_active(): bool {
+		$perfmatters_options = get_option( 'perfmatters_options' );
+
+		if ( ! empty( $perfmatters_options['assets']['remove_unused_css'] ) ) {
+			return true;
+		}
+
+		return false;
```
Comment (remyperona) [inline inc/ThirdParty/Plugins/Optimization/Perfmatters.php]:
```suggestion
		return empty( $perfmatters_options['assets']['remove_unused_css'] ) );
```

File: inc/ThirdParty/Plugins/Optimization/RapidLoad.php [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+	 *
+	 * @return array
+	 */
+	public static function get_subscribed_events() {
+		if ( ! defined( 'UUCSS_VERSION' ) ) {
+			return [];
+		}
+
+		return [
+			'rocket_maybe_disable_rucss'              => 'disable_rucss_option',
```
Comment (remyperona) [inline inc/ThirdParty/Plugins/Optimization/RapidLoad.php]:
```suggestion
			'rocket_disable_rucss_setting'              => 'disable_rucss_setting',
```

File: inc/ThirdParty/Plugins/Optimization/RapidLoad.php [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+			'pre_get_rocket_option_remove_unused_css' => 'maybe_disable_rucss',
+		];
+	}
+
+	/**
+	 * Disable RUCSS option.
```
Comment (remyperona) [inline inc/ThirdParty/Plugins/Optimization/RapidLoad.php]:
```suggestion
	 * Disable RUCSS setting.
```

File: inc/ThirdParty/Plugins/Optimization/RapidLoad.php [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+	 *
+	 * @param array $status RUCSS option status.
+	 * @return array
+	 */
+	public function disable_rucss_option( array $status ): array {
```
Comment (remyperona) [inline inc/ThirdParty/Plugins/Optimization/RapidLoad.php]:
```suggestion
	public function disable_rucss_setting( array $status ): array {
```

**Review by remyperona — CHANGES_REQUESTED** [review]

**Review by remyperona — APPROVED** [review]

**Comment by vmanthos** [comment]
With the latest changes, we are displaying the Rapidload notice when that's actually activated. Disabling the plugin enables our Remove Unused CSS back, however, that's not applied until the cache is cleared.

@piotrbak Do you think we should clear WP Rocket's cache when:
1. Rapidload's used CSS was enabled `AND`
2. The plugin was deactivated?

**Comment by piotrbak** [comment]
@jeawhanlee Vasilis is right there about the cache clearing. We should handle that in this PR to improve UX. Here is the scenario (thanks @vmanthos):

1. Our Remove Unused CSS is enabled.
2. They install Rapidload and enable their RUCSS.
3. We display the notice and disable ours.
4. **They either disable RUCSS or deactivate their plugin.**
5. Our RUCSS is automatically enabled.
6. We clear the cache for that to be applied to the front-end.

**Review by CrochetFeve0251 — APPROVED** [review]

**Comment by vmanthos** [comment]
@jeawhanlee  The trial of Rapidload has ended on my [test site](https://vasilis.rocketlabsqa.ovh/), and I'm getting this in "Settings > Rapidload":
image

Although the license information is "INVALID" our Remove Unused CSS is still disabled, the notice is displayed, and I cannot manually enable it.

Can you please look into this?

**Review by remyperona — APPROVED** [review]

**Review by vmanthos — APPROVED** [review]
Working as expected! :+1:

**Pull request merged**